src/compiler/scala/tools/nsc/plugins/Plugin.scala

```diff
@@ -155,40 +157,28 @@ object Plugin {
     paths: List[List[Path]],
     dirs: List[Path],
     ignoring: List[String],
-    disableClassLoaderCache: Boolean): List[Try[AnyClass]] =
+    findPluginClassloader: (Seq[Path] => ClassLoader)): List[Try[AnyClass]] =
   {
-    // List[(jar, Try(descriptor))] in dir
-    def scan(d: Directory) =
-      d.files.toList sortBy (_.name) filter (Jar isJarOrZip _) map (j => (j, loadDescriptionFromJar(j)))
-
     type PDResults = List[Try[(PluginDescription, ScalaClassLoader)]]
 
-    // scan plugin dirs for jars containing plugins, ignoring dirs with none and other jars
-    val fromDirs: PDResults = dirs filter (_.isDirectory) flatMap { d =>
-      scan(d.toDirectory) collect {
-        case (j, Success(pd)) => Success((pd, loaderFor(Seq(j), disableClassLoaderCache)))
+    val fromLoaders = paths.map {path =>
+      val loader = findPluginClassloader(path)
+      val list: Option[URL] = loader.getResources(PluginXML).asScala.take(1).toList.headOption
+      list match {
+        case Some(url) =>
+          val inputStream = url.openStream
+          try {
+            Try((PluginDescription.fromXML(inputStream), loader))
+          } finally {
+            inputStream.close()
+          }
+        case None =>
+          Failure(new MissingPluginException(path))
       }
```

Comment:
@som-snytt I'm thinking of making this change to read `scalac-plugin.xml` from a classloader, rather than by looking through JARs and directories. The motivation is to let customized `Global`-s provide the classloader, to do things like manage the lifecycle of the classloader + open file handles with extra knowledge of multi-project builds.

Seems to me like a logical extension of the groundwork you laid in scala/scala#3169. WDYT?

Reply:
Sounds good. I don't remember why the old logic looks so labored. Please indulge my alliteration.

By coincidence, at the day job I just asked to switch some legacy config from a command line option to a class loader resource.

I read the related tickets around class loader caching.